### PR TITLE
[lte][agw] Adding gtp statistics collector to pipelined

### LINKED ISSUE #2908: Add datapath metrics collection for eNodeb

### PR DESCRIPTION
**Pull request by ardzoht**
Signed-off-by: Alejandro Rodriguez <alexrod@fb.com>

## Summary

- Adds GTP stats collectors job to pipelined service
- Adds GTP_PORT_USER_PLANE_UL_BYTES. GTP_PORT_USER_PLANE_DL_BYTES counter metrics to pipelined (Per eNB datapath usage stats)

## Test Plan

- make test
- make integ_test
- Running local AGW with s1ap tests (setting `ovs_multi_tunnel: true` on spgw.yml) to activate dynamic GTP tunnel port creation and monitor metrics are registered

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+import logging
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py:14]: missing copyright

File: lte/gateway/python/magma/pipelined/main.py [hunk 2]
```diff
@@ -124,6 +125,11 @@ def on_exit_server_thread():
         start_check_quota_server(run_flask, bridge_ip, no_quota_port, False,
                                  on_exit_server_thread)
 
+    collector = GTPStatsCollector(
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/main.py]: Check that service is lte
    if service.config['setup_type'] == 'LTE':

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            self._loop,
+        )
+
+    async def _run(self) -> None:
+        logging.info("Running GTP stats collector...")
+        while True:
+            dump_stats_results = await self._ovsdb_dump_async(
+                'Interface', ['name', 'statistics', 'options'])
+            for r in list(dump_stats_results)[0].out:
+                if 'g_' in r.Interface:
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py]: [nit] maybe pull out `g_` as a const INTERFACE_PREFIX

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+])
+OVSDBCommandResult = namedtuple('OVSDBCommandResult', ['out', 'err'])
+
+interface_tx_rx_stats_re = re.compile(
+    r'"(?P<Interface>\w+)"(.*)remote_ip="(?P<remote_ip>.*)"(.*)rx_bytes=(?P<rx_bytes>\d).+tx_bytes=(?P<tx_bytes>\d)')
+
+
+class GTPStatsCollector(Job):
+    def __init__(self, polling_interval: int, service_loop):
+        self._polling_interval = max(polling_interval, 60)
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py]: [nit] maybe pull out 60 as a const MIN_INTERFACE_TIME

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        line_matches = []
+        for line in lines:
+            line_match = line_re.match(line)
+            if line_match:
+                line_matches.append(line_match.groupdict())
+        return line_matches
+
+    if stderr:
+        return create_error_result(stderr)
+    else:
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py]: maybe remove the else?

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        try:
+            stdout_lines = stdout.decode('ascii').split('\n')
+            header_line_idx = find_header_line_idx(stdout_lines)
+            gtp_matches = match_gtp_lines(
+                stdout_lines[header_line_idx + 1:],
+                interface_tx_rx_stats_re)
+
+            results = []
+            for m in gtp_matches:
+                logging.info(m)
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py]: Will this be too verbose? Maybe debug is better

**Review by koolzz — CHANGES_REQUESTED** [review]
Great stuff, added a few comments

**Review by koolzz — APPROVED** [review]
noice

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+ParsedInterfaceStats = namedtuple('ParsedInterfaceStats', [
+    'Interface',
+    'rx_bytes',
+    'tx_bytes',
+    'remote_ip',
+])
+OVSDBCommandResult = namedtuple('OVSDBCommandResult', ['out', 'err'])
+
+interface_tx_rx_stats_re = re.compile(
+    r'"(?P<Interface>\w+)"(.*)remote_ip="(?P<remote_ip>.*)"(.*)rx_bytes=(?P<rx_bytes>\d).+tx_bytes=(?P<tx_bytes>\d)')
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py]: remote_ip could be optional token, so that we can also report stats for gtp0 interface.

**Review by pshelar — CHANGES_REQUESTED** [review]

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py (hunk 7 above)
Comment (ardzoht) [reply lte/gateway/python/magma/pipelined/gtp_stats_collector.py]: Updated regex to include remote_ip as optional.

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    GTP_PORT_USER_PLANE_UL_BYTES
+
+OVSDBDumpCommandParams = namedtuple('OVSDBCommandParams', ['table', 'columns'])
+ParsedInterfaceStats = namedtuple('ParsedInterfaceStats', [
+    'Interface',
+    'rx_bytes',
+    'tx_bytes',
+    'remote_ip',
+])
+OVSDBCommandResult = namedtuple('OVSDBCommandResult', ['out', 'err'])
```
Comment (xjtian) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py]: use `typing.namedtuple` instead for these

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+interface_tx_rx_stats_re = re.compile(
+    r'"(?P<Interface>\w+)"(.*)(remote_ip="(?P<remote_ip>.*)"(.*))?rx_bytes=(?P<rx_bytes>\d).+tx_bytes=(?P<tx_bytes>\d)')
```
Comment (xjtian) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py]: for maintainability, can you build this up from a couple of intermediate strings, e.g.

```
iface_match = '"(?P<Interface>\w+)"'
remote_ip_match = '(remote_ip="(?P<remote_ip>)"'
# etc

final_re = re.compile(f'{iface_match}(.*){remote_ip_match}(.*)...')
```

Comment (xjtian) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py]: `\d` matches only 1 digit, is this on purpose, or do you mean `\d+`?

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+class GTPStatsCollector(Job):
+    def __init__(self, polling_interval: int, service_loop):
+        self._polling_interval = max(polling_interval,
+                                     MIN_OVSDB_DUMP_POLLING_INTERVAL)
+        super().__init__(interval=self._polling_interval, loop=service_loop)
+        self._loop = service_loop
+
+    @asyncio.coroutine
+    def _ovsdb_dump_async(self, table: str, columns: List[str]):
```
Comment (xjtian) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py:63]: async def?

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        """
+        params = [OVSDBDumpCommandParams(table=table, columns=columns)]
+        return subprocess_workflow.exec_and_parse_subprocesses_async(
+            params,
+            _get_ovsdb_dump_params,
+            _parse_ovsdb_dump_output,
+            self._loop,
+        )
+
+    async def _run(self) -> None:
```
Comment (xjtian) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py:77]: The Job superclass takes care of the periodicity already, you just need to run your task once in `_run` (get rid of the while)

**Review by xjtian — CHANGES_REQUESTED** [review]

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py (hunk 10 above)
Comment (ardzoht) [reply lte/gateway/python/magma/pipelined/gtp_stats_collector.py:63]: I tried async def but the issue is that  the subprocess_workflow call complains about the notation with `yield from`:

```
"/home/vagrant/magma/orc8r/gateway/python/magma/magmad/check/subprocess_workflow.py", line 78, in exec_and_parse_subprocesses_async
Oct 06 08:19:56 magma-dev pipelined[28161]:     subprocs = yield from asyncio.gather(*futures)
Oct 06 08:19:56 magma-dev pipelined[28161]:   File "/usr/lib/python3.5/asyncio/futures.py", line 381, in __iter__
Oct 06 08:19:56 magma-dev pipelined[28161]:     assert self.done(), "yield from wasn't used with future"
Oct 06 08:19:56 magma-dev pipelined[28161]: AssertionError: yield from wasn't used with future
```

File: lte/gateway/python/magma/pipelined/gtp_stats_collector.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+            'Interface', ['name', 'statistics', 'options'])
+        for r in list(dump_stats_results)[0].out:
+            if GTP_IP_INTERFACE_PREFIX in r.Interface or \
+                    r.Interface == GTP_INTERFACE_PREFIX:
+                logging.info(r.Interface)
+                GTP_PORT_USER_PLANE_DL_BYTES.labels(r.remote_ip).inc(
+                    float(r.rx_bytes))
+                GTP_PORT_USER_PLANE_UL_BYTES.labels(r.remote_ip).inc(
+                    float(r.tx_bytes))
+        await asyncio.sleep(self._polling_interval, self._loop)
```
Comment (xjtian) [inline lte/gateway/python/magma/pipelined/gtp_stats_collector.py]: superclass does the sleep

**Review by xjtian — APPROVED** [review]

**Review by pshelar — APPROVED** [review]

**Comment by xjtian** [comment]
cherry-picked onto v1.3

**Pull request merged**